File: app/src/main/java/com/mehmetakiftutuncu/muezzin/services/WidgetUpdaterService.java
package com.mehmetakiftutuncu.muezzin.services;

import android.app.IntentService;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.core.content.ContextCompat;
import android.view.View;
import android.widget.RemoteViews;

import com.github.mehmetakiftutuncu.toolbelt.Log;
import com.github.mehmetakiftutuncu.toolbelt.Optional;
import com.mehmetakiftutuncu.muezzin.R;
import com.mehmetakiftutuncu.muezzin.activities.PrayerTimesActivity;
import com.mehmetakiftutuncu.muezzin.models.Place;
import com.mehmetakiftutuncu.muezzin.models.PrayerTimesOfDay;
import com.mehmetakiftutuncu.muezzin.models.WidgetType;
import com.mehmetakiftutuncu.muezzin.utilities.Pref;
import com.mehmetakiftutuncu.muezzin.utilities.RemainingTime;
import com.mehmetakiftutuncu.muezzin.widgetproviders.PrayerTimesWidgetBase;

import org.joda.time.LocalTime;

/**
 * Created by akif on 08/07/16.
 */
public class WidgetUpdaterService extends IntentService {
    public WidgetUpdaterService() {
        super("WidgetUpdaterService");
    }

    public static void startServiceToUpdateWidget(Context context, WidgetType widgetType, int widgetId) {
        Intent intent = new Intent(context, WidgetUpdaterService.class);
        intent.putExtra("widgetType", widgetType.ordinal()).putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId);

        context.startService(intent);
    }

    @Override protected void onHandleIntent(Intent intent) {
        if (intent != null) {
            Bundle extras         = intent.getExtras();
            WidgetType widgetType = WidgetType.values()[extras.getInt("widgetType", 0)];
            int widgetId          = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID);

            update(widgetType, widgetId);
        }
    }

    private void update(WidgetType widgetType, int widgetId) {
        Log.debug(getClass(), "Updating widget '%s' with id '%d'...", widgetType, widgetId);

        // Get RemoteViews and AppWidgetManager to work with
        RemoteViews remoteViews           = new RemoteViews(getPackageName(), widgetType.layoutId);
        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(this);

        // A pending intent to launch prayer times activity
        Intent intent               = new Intent(this, PrayerTimesActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, intent, 0);

        Optional<Place> maybePlace = Pref.Places.getCurrentPlace(this);

        Optional<PrayerTimesOfDay> maybePrayerTimesForToday = maybePlace.isDefined() ? PrayerTimesOfDay.getPrayerTimesForToday(this, maybePlace.get()) : Optional.empty();

        // When clicked on the root item in widget layout, which is basically anywhere on the widget, launch the pending intent
        remoteViews.setOnClickPendingIntent(R.id.linearLayout_widgetLayout, pendingIntent);

        // Initialize prayer times texts
        if (maybePlace.isDefined()) {
            Place place = maybePlace.get();

            if (maybePrayerTimesForToday.isDefined()) {
                PrayerTimesOfDay prayerTimes = maybePrayerTimesForToday.get();

                switch (widgetType) {
                    case PRAYER_TIMES_HORIZONTAL:
                        updatePlaceName(remoteViews, place, true);
                        updatePrayerTimes(remoteViews, prayerTimes);
                        break;

                    case PRAYER_TIMES_VERTICAL:
                        updatePlaceName(remoteViews, place, false);
                        updatePrayerTimes(remoteViews, prayerTimes);
                        break;

                    case PRAYER_TIMES_BIG:
                        updatePlaceName(remoteViews, place, true);
                        updatePrayerTimes(remoteViews, prayerTimes);
                        updateRemainingTime(remoteViews, prayerTimes);
                        break;
                }
            } else {
                PrayerTimesWidgetBase.cancelUpdates(this, widgetType);
            }
        } else {
            PrayerTimesWidgetBase.cancelUpdates(this, widgetType);
        }

        // Apply all these updates above to the widget
        appWidgetManager.updateAppWidget(widgetId, remoteViews);
    }

    private void updatePlaceName(RemoteViews remoteViews, Place place, boolean isFullNameRequired) {
        remoteViews.setTextViewText(R.id.textView_prayerTimes_placeName, place.getPlaceName(getBaseContext(), isFullNameRequired).getOrElse(getString(R.string.applicationName)));
    }

    private void updatePrayerTimes(RemoteViews remoteViews, PrayerTimesOfDay prayerTimes) {
        int defaultTextColor = ContextCompat.getColor(this, android.R.color.secondary_text_light);
        int greenTextColor   = ContextCompat.getColor(this, R.color.colorPrimary);
        int redTextColor     = ContextCompat.getColor(this, R.color.red);

        LocalTime nextPrayerTime = prayerTimes.nextPrayerTime();
        LocalTime remaining  = RemainingTime.to(nextPrayerTime);

        boolean isRemainingLessThan45Minutes = remaining.getHourOfDay() == 0 && remaining.getMinuteOfHour() < 45;

        remoteViews.setTextViewText(R.id.textView_prayerTimes_fajrTime,    prayerTimes.fajr.toString(PrayerTimesOfDay.TIME_FORMATTER));
        remoteViews.setTextViewText(R.id.textView_prayerTimes_shuruqTime,  prayerTimes.shuruq.toString(PrayerTimesOfDay.TIME_FORMATTER));
        remoteViews.setTextViewText(R.id.textView_prayerTimes_dhuhrTime,   prayerTimes.dhuhr.toString(PrayerTimesOfDay.TIME_FORMATTER));
        remoteViews.setTextViewText(R.id.textView_prayerTimes_asrTime,     prayerTimes.asr.toString(PrayerTimesOfDay.TIME_FORMATTER));
        remoteViews.setTextViewText(R.id.textView_prayerTimes_maghribTime, prayerTimes.maghrib.toString(PrayerTimesOfDay.TIME_FORMATTER));
        remoteViews.setTextViewText(R.id.textView_prayerTimes_ishaTime,    prayerTimes.isha.toString(PrayerTimesOfDay.TIME_FORMATTER));

        remoteViews.setTextColor(R.id.textView_prayerTimes_fajrTimeName,    defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_shuruqTimeName,  defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_dhuhrTimeName,   defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_asrTimeName,     defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_maghribTimeName, defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_ishaTimeName,    defaultTextColor);

        remoteViews.setTextColor(R.id.textView_prayerTimes_fajrTime,    defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_shuruqTime,  defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_dhuhrTime,   defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_asrTime,     defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_maghribTime, defaultTextColor);
        remoteViews.setTextColor(R.id.textView_prayerTimes_ishaTime,    defaultTextColor);

        int color = isRemainingLessThan45Minutes ? redTextColor : greenTextColor;

        switch (prayerTimes.nextPrayerTimeType()) {
            case Fajr:
                remoteViews.setTextColor(R.id.textView_prayerTimes_fajrTimeName, color);
                remoteViews.setTextColor(R.id.textView_prayerTimes_fajrTime, color);
                break;

            case Shuruq:
                remoteViews.setTextColor(R.id.textView_prayerTimes_shuruqTimeName, color);
                remoteViews.setTextColor(R.id.textView_prayerTimes_shuruqTime, color);
                break;

            case Dhuhr:
                remoteViews.setTextColor(R.id.textView_prayerTimes_dhuhrTimeName, color);
                remoteViews.setTextColor(R.id.textView_prayerTimes_dhuhrTime, color);
                break;

            case Asr:
                remoteViews.setTextColor(R.id.textView_prayerTimes_asrTimeName, color);
                remoteViews.setTextColor(R.id.textView_prayerTimes_asrTime, color);
                break;

            case Maghrib:
                remoteViews.setTextColor(R.id.textView_prayerTimes_maghribTimeName, color);
                remoteViews.setTextColor(R.id.textView_prayerTimes_maghribTime, color);
                break;

            case Isha:
                remoteViews.setTextColor(R.id.textView_prayerTimes_ishaTimeName, color);
                remoteViews.setTextColor(R.id.textView_prayerTimes_ishaTime, color);
                break;
        }
    }

    private void updateRemainingTime(RemoteViews remoteViews, PrayerTimesOfDay prayerTimes) {
        int defaultTextColor = ContextCompat.getColor(this, android.R.color.secondary_text_light);
        int redTextColor     = ContextCompat.getColor(this, R.color.red);

        LocalTime nextPrayerTime  = prayerTimes.nextPrayerTime();
        String nextPrayerTimeName = PrayerTimesOfDay.prayerTimeLocalizedName(this, prayerTimes.nextPrayerTimeType());

        LocalTime remaining  = RemainingTime.to(nextPrayerTime);
        String remainingTime = remaining.toString(RemainingTime.FORMATTER_WITHOUT_SECONDS);

        boolean isRemainingLessThan45Minutes = remaining.getHourOfDay() == 0 && remaining.getMinuteOfHour() < 45;

        int color = isRemainingLessThan45Minutes ? redTextColor : defaultTextColor;

        remoteViews.setViewVisibility(R.id.linearLayout_prayerTimes_remainingTime, View.VISIBLE);

        remoteViews.setTextViewText(R.id.textView_prayerTimes_remainingTimeInfo, getString(R.string.prayerTimes_cardTitle_remainingTime, nextPrayerTimeName));
        remoteViews.setTextViewText(R.id.textView_prayerTimes_remainingTime, remainingTime);

        remoteViews.setTextColor(R.id.textView_prayerTimes_remainingTimeInfo, color);
        remoteViews.setTextColor(R.id.textView_prayerTimes_remainingTime, color);
    }
}

File: app/src/main/java/com/mehmetakiftutuncu/muezzin/services/WidgetUpdaterService.kt
package com.mehmetakiftutuncu.muezzin.services

import android.app.IntentService
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.RemoteViews
import androidx.core.content.ContextCompat
import com.github.mehmetakiftutuncu.toolbelt.Log
import com.mehmetakiftutuncu.muezzin.R
import com.mehmetakiftutuncu.muezzin.activities.PrayerTimesActivity
import com.mehmetakiftutuncu.muezzin.models.Place
import com.mehmetakiftutuncu.muezzin.models.PrayerTimeType
import com.mehmetakiftutuncu.muezzin.models.PrayerTimesOfDay
import com.mehmetakiftutuncu.muezzin.models.WidgetType
import com.mehmetakiftutuncu.muezzin.repositories.PlaceRepository
import com.mehmetakiftutuncu.muezzin.repositories.PrayerTimesOfDayRepository
import com.mehmetakiftutuncu.muezzin.utilities.Pref
import com.mehmetakiftutuncu.muezzin.utilities.RemainingTime
import com.mehmetakiftutuncu.muezzin.widget.PrayerTimesWidget

class WidgetUpdaterService: IntentService("WidgetUpdaterService") {
    private val ctx: Context by lazy { this }

    override fun onHandleIntent(intent: Intent?) {
        intent?.extras?.also {
            update(
                WidgetType.values()[it.getInt("widgetType", 0)],
                it.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID)
            )
        }
    }

    private fun update(type: WidgetType, id: Int) {
        Log.debug(javaClass, "Updating widget '$type' with id '$id'...")

        // Get RemoteViews and AppWidgetManager to work with
        val remoteViews = RemoteViews(packageName, type.layoutId)
        val appWidgetManager = AppWidgetManager.getInstance(ctx)

        // A pending intent to launch prayer times activity
        val intent = Intent(ctx, PrayerTimesActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(ctx, 0, intent, 0)

        // When clicked on the root item in widget layout,
        // which is basically anywhere on the widget, launch the pending intent
        remoteViews.setOnClickPendingIntent(R.id.linearLayout_widgetLayout, pendingIntent)

        val place = Pref.Places.getCurrentPlace(ctx)

        if (place == null) {
            PrayerTimesWidget.cancelUpdates(ctx, type)
        } else {
            val times = PrayerTimesOfDayRepository.getForToday(ctx, place)

            if (times == null) {
                PrayerTimesWidget.cancelUpdates(ctx, type)
            } else {
                // Initialize prayer times texts
                when (type) {
                    WidgetType.Horizontal -> {
                        updatePlaceName(remoteViews, place, true)
                        updatePrayerTimes(remoteViews, times)
                    }

                    WidgetType.Vertical -> {
                        updatePlaceName(remoteViews, place, false)
                        updatePrayerTimes(remoteViews, times)
                    }

                    WidgetType.Big -> {
                        updatePlaceName(remoteViews, place, true)
                        updatePrayerTimes(remoteViews, times)
                        updateRemainingTime(remoteViews, times)
                    }
                }
            }
        }

        // Apply all these updates above to the widget
        appWidgetManager.updateAppWidget(id, remoteViews)
    }

    private fun updatePlaceName(remoteViews: RemoteViews, place: Place, fullNameRequired: Boolean) {
        PlaceRepository.getName(baseContext, place, fullNameRequired)?.let {
            remoteViews.setTextViewText(R.id.textView_prayerTimes_placeName, it)
        }
    }

    private fun updatePrayerTimes(remoteViews: RemoteViews, prayerTimes: PrayerTimesOfDay) {
        val defaultTextColor = ContextCompat.getColor(ctx, android.R.color.secondary_text_light)
        val greenTextColor = ContextCompat.getColor(ctx, R.color.colorPrimary)
        val redTextColor = ContextCompat.getColor(ctx, R.color.red)
        val nextPrayerTime = prayerTimes.nextPrayerTime()
        val remaining = RemainingTime.to(nextPrayerTime)
        val lessThan45Minutes = remaining.hourOfDay == 0 && remaining.minuteOfHour < 45

        remoteViews.apply {
            setTextViewText(R.id.textView_prayerTimes_fajrTime, prayerTimes.fajr.toString(PrayerTimesOfDay.timeFormatter))
            setTextViewText(R.id.textView_prayerTimes_shuruqTime, prayerTimes.shuruq.toString(PrayerTimesOfDay.timeFormatter))
            setTextViewText(R.id.textView_prayerTimes_dhuhrTime, prayerTimes.dhuhr.toString(PrayerTimesOfDay.timeFormatter))
            setTextViewText(R.id.textView_prayerTimes_asrTime, prayerTimes.asr.toString(PrayerTimesOfDay.timeFormatter))
            setTextViewText(R.id.textView_prayerTimes_maghribTime, prayerTimes.maghrib.toString(PrayerTimesOfDay.timeFormatter))
            setTextViewText(R.id.textView_prayerTimes_ishaTime, prayerTimes.isha.toString(PrayerTimesOfDay.timeFormatter))

            setTextColor(R.id.textView_prayerTimes_fajrTimeName, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_shuruqTimeName, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_dhuhrTimeName, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_asrTimeName, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_maghribTimeName, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_ishaTimeName, defaultTextColor)

            setTextColor(R.id.textView_prayerTimes_fajrTime, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_shuruqTime, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_dhuhrTime, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_asrTime, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_maghribTime, defaultTextColor)
            setTextColor(R.id.textView_prayerTimes_ishaTime, defaultTextColor)

            val color = if (lessThan45Minutes) redTextColor else greenTextColor

            when (prayerTimes.nextPrayerTimeType()) {
                PrayerTimeType.Fajr -> {
                    setTextColor(R.id.textView_prayerTimes_fajrTimeName, color)
                    setTextColor(R.id.textView_prayerTimes_fajrTime, color)
                }

                PrayerTimeType.Shuruq -> {
                    setTextColor(R.id.textView_prayerTimes_shuruqTimeName, color)
                    setTextColor(R.id.textView_prayerTimes_shuruqTime, color)
                }

                PrayerTimeType.Dhuhr -> {
                    setTextColor(R.id.textView_prayerTimes_dhuhrTimeName, color)
                    setTextColor(R.id.textView_prayerTimes_dhuhrTime, color)
                }

                PrayerTimeType.Asr -> {
                    setTextColor(R.id.textView_prayerTimes_asrTimeName, color)
                    setTextColor(R.id.textView_prayerTimes_asrTime, color)
                }

                PrayerTimeType.Maghrib -> {
                    setTextColor(R.id.textView_prayerTimes_maghribTimeName, color)
                    setTextColor(R.id.textView_prayerTimes_maghribTime, color)
                }

                PrayerTimeType.Isha -> {
                    setTextColor(R.id.textView_prayerTimes_ishaTimeName, color)
                    setTextColor(R.id.textView_prayerTimes_ishaTime, color)
                }

                else -> {}
            }
        }
    }

    private fun updateRemainingTime(remoteViews: RemoteViews, prayerTimes: PrayerTimesOfDay) {
        val defaultTextColor = ContextCompat.getColor(ctx, android.R.color.secondary_text_light)
        val redTextColor = ContextCompat.getColor(ctx, R.color.red)
        val nextPrayerTime = prayerTimes.nextPrayerTime()
        val nextPrayerTimeName = PrayerTimesOfDay.localizedName(ctx, prayerTimes.nextPrayerTimeType())
        val remaining = RemainingTime.to(nextPrayerTime)
        val remainingTime = remaining.toString(RemainingTime.formatterWithoutSeconds)
        val lessThan45Minutes = remaining.hourOfDay == 0 && remaining.minuteOfHour < 45

        val color = if (lessThan45Minutes) redTextColor else defaultTextColor

        val remainingTimeInfo = getString(R.string.prayerTimes_cardTitle_remainingTime, nextPrayerTimeName)

        remoteViews.apply {
            setViewVisibility(R.id.linearLayout_prayerTimes_remainingTime, View.VISIBLE)
            setTextViewText(R.id.textView_prayerTimes_remainingTimeInfo, remainingTimeInfo)
            setTextViewText(R.id.textView_prayerTimes_remainingTime, remainingTime)
            setTextColor(R.id.textView_prayerTimes_remainingTimeInfo, color)
            setTextColor(R.id.textView_prayerTimes_remainingTime, color)
        }
    }

    companion object {
        fun start(ctx: Context, type: WidgetType, id: Int) {
            Intent(ctx, WidgetUpdaterService::class.java).apply {
                putExtra("widgetType", type.ordinal)
                putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, id)
            }.also {
                ctx.startService(it)
            }
        }
    }
}